tsdb/meta.go

```diff
@@ -1423,40 +1441,55 @@ func (a Measurements) union(other Measurements) Measurements {
 type Series struct {
 	mu          sync.RWMutex
 	Key         string
-	Tags        map[string]string
+	Tags        models.Tags
 	id          uint64
 	measurement *Measurement
-	shardIDs    map[uint64]bool // shards that have this series defined
+	shardIDs    []uint64 // shards that have this series defined
 }
 
 // NewSeries returns an initialized series struct
-func NewSeries(key string, tags map[string]string) *Series {
+func NewSeries(key string, tags models.Tags) *Series {
 	return &Series{
-		Key:      key,
-		Tags:     tags,
-		shardIDs: make(map[uint64]bool),
+		Key:  key,
+		Tags: tags,
 	}
 }
 
 func (s *Series) AssignShard(shardID uint64) {
 	s.mu.Lock()
-	s.shardIDs[shardID] = true
+	if !s.assigned(shardID) {
+		s.shardIDs = append(s.shardIDs, shardID)
+		sort.Sort(uint64Slice(s.shardIDs))
+	}
 	s.mu.Unlock()
 }
 
 func (s *Series) UnassignShard(shardID uint64) {
 	s.mu.Lock()
-	delete(s.shardIDs, shardID)
+	for i, v := range s.shardIDs {
+		if v == shardID {
+			s.shardIDs = append(s.shardIDs[:i], s.shardIDs[i+1:]...)
+		}
+	}
 	s.mu.Unlock()
 }
 
 func (s *Series) Assigned(shardID uint64) bool {
 	s.mu.RLock()
-	b := s.shardIDs[shardID]
+	b := s.assigned(shardID)
 	s.mu.RUnlock()
 	return b
 }
 
+func (s *Series) assigned(shardID uint64) bool {
+	for _, v := range s.shardIDs {
+		if v == shardID {
```

Comment:
Changed to `sort.Search` in 3acdf01. We could potentially have a large number of shards assigned to a series so making it searchable makes sense.